### PR TITLE
Documented System.Text.Unicode namespace

### LINKED ISSUE #2983: Document System.​Text.​Unicode Namespace
This includes the following files:

https://github.com/dotnet/docs/blob/master/xml/ns-System.Text.Unicode.xml 
https://github.com/dotnet/docs/blob/master/xml/System.Text.Unicode/UnicodeRange.xml
https://github.com/dotnet/docs/blob/master/xml/System.Text.Unicode/UnicodeRanges.xml

### PR DESCRIPTION
**Pull request by rpetrusha**
# Documented System.Text.Unicode namespace

## Summary

Documents the **System.Text.Unicode** namespace, as well as the **UnicodeRange** and **UnicodeRanges** types.

Fixes #2983 

## Suggested Reviewers

@mairaw, @tarekgh , @KrzysztofCwalina

### CODE REVIEW

File: xml/System.Text.Unicode/UnicodeRange.xml [hunk 1]
```diff
@@ -33,10 +39,21 @@
...
+        <param name="firstCodePoint">The first code point in the range.</param>
+        <param name="length">The number of code points in the range.</param>
+        <summary>Creates a new <see cref="System.Text.Unicode.UnicodeRange" /> that includes a specified number of characters starting at a specified Unicode code point.</summary>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException"> 
+           <paramref name="firstCodePoint" /> lies outside of the Basic Multilingual Plane. 
+
+-or-           
+
+<paramref name="length" /> is less than zero.
```
Comment (tarekgh) [inline xml/System.Text.Unicode/UnicodeRange.xml]: >is less than zero. [](start = 27, length = 18)

or the length will get the range beyond 0xFFFF

File: xml/System.Text.Unicode/UnicodeRanges.xml [hunk 2]
```diff
@@ -13,8 +13,16 @@
   </Base>
   <Interfaces />
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Provides static methods that return predefined <see cref="System.Text.Unicode.UnicodeRange" /> instances that correspond to blocks from the Unicode 7.0 specification.</summary>
```
Comment (tarekgh) [inline xml/System.Text.Unicode/UnicodeRanges.xml]: >7.0  [](start = 161, length = 4)

I am not sure if we need to specify the Unicode version here. we can just omit the version to have more flexibility and don't have to change the documentation later.

File: xml/System.Text.Unicode/UnicodeRanges.xml [hunk 3]
```diff
@@ -3172,8 +3180,8 @@
         <ReturnType>System.Text.Unicode.UnicodeRange</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
-        <value>To be added.</value>
+        <summary>Gets the Yi Syllables Unicode block (U+A000-U+A48C).</summary>
```
Comment (tarekgh) [inline xml/System.Text.Unicode/UnicodeRanges.xml]: >A48C [](start = 63, length = 4)

A48F?

File: xml/System.Text.Unicode/UnicodeRanges.xml [hunk 4]
```diff
@@ -3152,8 +3160,8 @@
         <ReturnType>System.Text.Unicode.UnicodeRange</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>To be added.</summary>
-        <value>To be added.</value>
+        <summary>Gets the Yi Radicals Unicode block (U+A490-U+A4C6).</summary>
```
Comment (tarekgh) [inline xml/System.Text.Unicode/UnicodeRanges.xml]: >A4C6 [](start = 62, length = 4)

A4CF?

File: xml/System.Text.Unicode/UnicodeRanges.xml (hunk 3 above)
Comment (tarekgh) [reply xml/System.Text.Unicode/UnicodeRanges.xml]: I suggest you get the ranges from the file [https://raw.githubusercontent.com/dotnet/corefx/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UnicodeRanges.generated.cs](https://raw.githubusercontent.com/dotnet/corefx/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UnicodeRanges.generated.cs)

---
In reply to: [140909609](https://github.com/dotnet/docs/pull/3223#discussion_r140909609) [](ancestors = 140909609)

**Comment by rpetrusha** [comment]
@tarekgh, as the end code point for a block, I used the last valid code point in that block, and not the last assigned code point. I think I should revise this; otherwise, going back to your comment on the Unicode standard version, we'll have to revise this each time a new version of the Unicode standard is released.

**Comment by tarekgh** [comment]
>  going back to your comment on the Unicode standard version, we'll have to revise this each time a new version of the Unicode standard is released.

My only concern with that every time we touch the ranges we have to remember to update this version. I am not sure how we can guarantee the sync between the data we have and the referenced version.

**Comment by rpetrusha** [comment]
@tarekgh, I've revised the document to address your suggestions, as well as to correct some typos and other errors I found. In general, I agree that we should document the beginning and end of ranges as specified by the Unicode standard without excluding undefined code points at the start or end of the range; you're right that we can't realistically maintain the documentation whenever the supported Unicode version changes. So I've done that for all of the fields in the UnicodeRanges class.

File: xml/System.Text.Unicode/UnicodeRange.xml (hunk 1 above)
Comment (tarekgh) [reply xml/System.Text.Unicode/UnicodeRange.xml]: I am not seeing you have added the suggested part

---
In reply to: [140906341](https://github.com/dotnet/docs/pull/3223#discussion_r140906341) [](ancestors = 140906341)

File: xml/System.Text.Unicode/UnicodeRange.xml [hunk 5]
```diff
@@ -33,10 +42,25 @@
...
+        <exception cref="T:System.ArgumentOutOfRangeException"> 
+
+           <paramref name="firstCodePoint" /> is less than zero. 
+
+-or-
+           <paramref name="firstCodePoint" /> is greater than 0xFFFF. 
+
+-or-           
+
+<paramref name="length" /> is less than zero.
```
Comment (tarekgh) [inline xml/System.Text.Unicode/UnicodeRange.xml]: We need to add if the length will get us outside the oxFFFF range too.

**Review by tarekgh — APPROVED** [review]

File: xml/System.Text.Unicode/UnicodeRange.xml (hunk 5 above)
Comment (rpetrusha) [reply xml/System.Text.Unicode/UnicodeRange.xml]: @tarekgh, I've added it.

**Comment by rpetrusha** [comment]
Closing and reopening to begin new build.

File: xml/System.Text.Unicode/UnicodeRange.xml [hunk 6]
```diff
@@ -13,8 +13,17 @@
...
   <Interfaces />
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Represents a contiguous range of Unicode code points.</summary>
+    <remarks>
+      <format type="text/markdown"><![CDATA[  
+## Remarks  
+
+You use the static <xref:System.Text.Unicode.UnicodeRange.Create%2A> method or the <xref:System.Text.Unicode.UnicodeRange.#ctor%2A> constructor to create a an arbitrary range of Unicode code points. You use the static properties of the <xref:System.Text.Unicode.UnicodeRanges> class retrieve a <xref:System.Text.Unicode.UnicodeRange> instance that contains the code points in a Unicode named range.
```
Comment (mairaw) [inline xml/System.Text.Unicode/UnicodeRange.xml]: the #ctor xref link is broken

File: xml/System.Text.Unicode/UnicodeRange.xml [hunk 7]
```diff
@@ -33,10 +42,25 @@
...
+        <param name="firstCodePoint">The first code point in the range.</param>
+        <param name="length">The number of code points in the range.</param>
+        <summary>Creates a new <see cref="System.Text.Unicode.UnicodeRange" /> that includes a specified number of characters starting at a specified Unicode code point.</summary>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException"> 
+
+           <paramref name="firstCodePoint" /> is less than zero or greater than 0xFFFF. 
+
+-or-
+           <paramref name="firstCodePoint" /> is greater than 0xFFFF.
```
Comment (mairaw) [inline xml/System.Text.Unicode/UnicodeRange.xml]: it needs a new line here or it renders together with the next sentence 😢

File: xml/System.Text.Unicode/UnicodeRange.xml [hunk 8]
```diff
@@ -58,11 +82,13 @@
...
       </Parameters>
       <Docs>
-        <param name="firstCharacter">To be added.</param>
-        <param name="lastCharacter">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="firstCharacter">The first character in the range.</param>
+        <param name="lastCharacter">The last character in the range.</param>
+        <summary>Creates a new <see cref="System.Text.Unicode.UnicodeRange" /> instance from a span of characters.</summary>
+        <returns>A range that includes all characters between <param name="firstCharacter"/> and <param name="lastCharacter"/>.</returns>
```
Comment (mairaw) [inline xml/System.Text.Unicode/UnicodeRange.xml]: this is not rendering. need to check if this is the right tag for parameters

File: xml/System.Text.Unicode/UnicodeRange.xml [hunk 9]
```diff
@@ -58,11 +82,13 @@
...
         <remarks>To be added.</remarks>
+        <exception cref="System.ArgumentOutOfRangeException"><param name="lastCharacter"/> precedes <param name="firstCharacter" />.
```
Comment (mairaw) [inline xml/System.Text.Unicode/UnicodeRange.xml]: ditto

**Review by mairaw — APPROVED** [review]
LGTM. A few rendering issues to fix before merging.

**Comment by mairaw** [comment]
Curious, do you think the overload link doesn't work when we don't have an overload page? That seems like a bug to me.

**Review by mairaw — APPROVED** [review]
LGTM

**Comment by rpetrusha** [comment]
@mairaw, it turns out that the overload link doesn't work because the `#` in `#ctor` needs to be encoded.

**Pull request merged**